cody-repo-util.el

```diff
@@ -0,0 +1,182 @@
+;;; cody-repo-util.el --- Sourcegraph Cody repository utilities -*- lexical-binding: t; -*-
+
+;; Copyright (C) 2024 Sourcegraph, Inc.
+
+;; Version: 0.1
+;; Maintainer: Steve Yegge <stevey@sourcegraph.com>
+;; URL: https://github.com/sourcegraph/emacs-cody
+;; Package-Requires: ((emacs "26.3") (cl-lib "0.5"))
+
+;;; Commentary:
+;;
+;; This package provides utility functions to interact with version control
+;; systems within Sourcegraph Cody for Emacs.
+;;
+;; Functions in this package are used to retrieve repository information,
+;; handle URL replacements, and determine the VCS type for files in a project.
+
+;;; Code:
+
+(require 'cl-lib)
+(require 'vc)
+(require 'vc-git)
+(require 'url-parse)
+(require 'url-util) ; For `url-hexify-string`
+
+(defun cody--uri-for (buffer-or-file)
+  "Convert the BUFFER-OR-FILE to a URI string.
+Handles Windows drive letters and formats them appropriately.
+If BUFFER-OR-FILE is a buffer, retrieves its associated file name.
+Returns BUFFER-OR-FILE in case of errors or if conversion is not possible,
+or `buffer-name` as a last resort, to avoid sending a nil path to the agent."
+  (let* ((buffer (if (bufferp buffer-or-file) buffer-or-file (current-buffer)))
+         (file (or (and (bufferp buffer-or-file) (buffer-file-name buffer-or-file))
+                   (when (stringp buffer-or-file) buffer-or-file)
+                   (buffer-name buffer))))
+    (condition-case err
+        (let* ((expanded (if (string-match "^[a-zA-Z]:" file)
+                             file
+                           (expand-file-name file)))
+               (normalized (replace-regexp-in-string "\\\\" "/" expanded)) ; Normalize Windows paths
+               (uri (if (file-remote-p normalized)
+                        normalized
+                      (concat "file://"
+                              (if (string-match "\\([a-zA-Z]\\):" normalized)
+                                  ;; Handle Windows drive letters
+                                  (concat "/" (downcase (match-string 1 normalized)) "%3A"
+                                          (substring normalized 2))
+                                normalized)))))
+          uri)
+      (error
+       (message "Error converting file to URI: %s" err)
+       (buffer-name buffer)))))
+
+(defun cody--repo-info (project file)
+  "Return repository info for PROJECT and FILE."
+  (let* ((vcs-type (cody--get-vcs-type project file))
+         (relative-path "")
+         (remote-url "")
+         (remote-branch-name "")
+         (repo-root-path (cody--get-repo-root-path project file)))
+    (when repo-root-path
+      (setq relative-path (if (> (length (file-relative-name file repo-root-path)) 0)
```

Comment:
When is `file-relative-name` ever 0 length? Outside the repo it's going to be some number of `../` and `file`; inside the directory it is going to be `file`, except if `file` is `""` and then we have `.`; etc.